cmd/ingest/batch.go

```diff
@@ -0,0 +1,199 @@
+		}
+		p.doneCh <- struct{}{}
+	}()
+
+}
+
+func (p *BatchProcessor) Wait() {
+	fmt.Println("deleteme waiting 1")
+	p.incomingWg.Wait()
+	close(p.incomingCh)
+	fmt.Println("deleteme waiting 2")
+	<-p.doneCh
+	fmt.Println("deleteme waiting 3")
+	fmt.Printf("# reschedules: %d\n", p.reschedules.Load())
+}
+
+// Process processes a Batch into the DB.
+func (p *BatchProcessor) Process(b *Batch) {
+	p.incomingWg.Add(1) // one more batch to process
+	go func() {
+		p.incomingCh <- b
+	}()
+}
+
+// wrapBatch protects the processing of a batch.
+func (p *BatchProcessor) wrapBatch(batch *Batch) {
+	if err := p.checkIfBatchClashes(batch); err != nil {
+		// At least one name in this batch is already being processed at a different batch,
+		// and we can't use different batches that contain a non nil intersection.
+		// Just reschedule the batch in the hopes that it will eventually picked up when
+		// the active batches don't clash with it:
+		p.reschedules.Add(1)
+		p.incomingCh <- batch
+		return
+	}
+
+	p.addBatchAsActive(batch)
+	defer p.removeBatchFromActive(batch)
+	defer p.incomingWg.Done() // one less batch to process
+
+	p.processBatch(batch)
+	fmt.Println("batch processed")
+}
+
+func (p *BatchProcessor) processBatch(batch *Batch) {
+	// Compute which domains could be affected:
+	affectedDomainsMap, domainCertMap, domainCertChainMap := updater.GetAffectedDomainAndCertMap(
+		batch.Certs, batch.Chains)
+	if len(affectedDomainsMap) == 0 {
+		return
+	}
+
+	// Get all affected entries already present in the DB:
+	affectedDomainHashes := make([]common.SHA256Output, 0, len(affectedDomainsMap))
+	for k := range affectedDomainsMap {
+		affectedDomainHashes = append(affectedDomainHashes, k)
+	}
+	domainEntries, err := p.conn.RetrieveDomainEntries(context.Background(), affectedDomainHashes)
+	if err != nil {
+		panic(err)
+	}
+
+	// Obtain a map from SHAs to certificates:
+	shaToCerts := make(map[common.SHA256Output]*mcommon.DomainEntry)
+	for _, kv := range domainEntries {
+		entry, err := mcommon.DeserializeDomainEntry(kv.Value)
+		if err != nil {
+			panic(err)
+		}
+		shaToCerts[kv.Key] = entry
+	}
+
+	// Update Domain Entries in DB:
+	updatedDomains, err := updater.UpdateDomainEntries(shaToCerts, domainCertMap, domainCertChainMap)
+	if err != nil {
+		panic(err)
+	}
+	shaToCerts, err = updater.GetDomainEntriesToWrite(updatedDomains, shaToCerts)
+	if err != nil {
+		panic(err)
+	}
+	domainEntries, err = updater.SerializeUpdatedDomainEntries(shaToCerts)
+	if err != nil {
+		panic(err)
+	}
+	_, err = p.conn.UpdateDomainEntries(context.Background(), domainEntries)
+	if err != nil {
+		panic(err)
+	}
+
+	inputKeys, inputValues, err := updater.KeyValuePairToSMTInput(domainEntries)
+	if err != nil {
+		panic(err)
+	}
+	// TODO(juagargi) update SMT with the above
+	_ = inputKeys
+	_ = inputValues
+}
+
+func (p *BatchProcessor) checkIfBatchClashes(b *Batch) error {
+	p.runningBatchesMu.Lock()
+	defer p.runningBatchesMu.Unlock()
+
+	for _, cn := range b.cns {
```

Comment:
We are currently ignoring SAN entries (see comments above), which means there could be undetected clashes between batches (because the insertion takes SAN entries into account, see line 117 which calls `GetAffectedDomainAndCertMap`. This should become a nested loop once the above is fixed.